influxql/query_executor.go

```diff
@@ -341,52 +341,63 @@ func (e *QueryExecutor) query(qid uint64) (*QueryTask, bool) {
 	return query, ok
 }
 
-// attachQuery attaches a running query to be managed by the QueryExecutor.
-// Returns the query id of the newly attached query or an error if it was
-// unable to assign a query id or attach the query to the QueryExecutor.
-// This function also returns a channel that will be closed when this
-// query finishes running.
-//
-// After a query finishes running, the system is free to reuse a query id.
-func (e *QueryExecutor) attachQuery(q *Query, database string, interrupt <-chan struct{}) (uint64, *QueryTask, error) {
-	e.mu.Lock()
-	defer e.mu.Unlock()
+// newQuery creates a new query to be managed by the QueryExecutor.
+func (e *QueryExecutor) newQuery(q *Query, database string, interrupt <-chan struct{}) (uint64, *QueryTask, error) {
+	task := &QueryTask{
+		query:     q.String(),
+		database:  database,
+		startTime: time.Now(),
+		closing:   make(chan struct{}),
+		monitorCh: make(chan error),
+	}
 
+	e.mu.Lock()
```

Comment:
Are you sure that this actually reduces the lock contention though?  The lock contention we saw was on trying to acquire the write lock.  We usually have had to use a read-lock and then re-acquire/check under a write-lock if necessary to reduce write lock contention.  For example: https://github.com/influxdata/influxdb/blob/master/tsdb/engine/tsm1/cache.go#L458